text/0000-reactivity-update.md

```diff
@@ -0,0 +1,327 @@
+
+Reasoning:
+
+- Doesn't rely on counters with magic numbers
+- Quicky stops executing on possible loops
+- The reactive statement will run eventually (when another change is detected and scheduled)
+- Not too complicated to implement
+- Changes of not working correctly much lower than the current implementation
+
+## Future improvements ideas:
+
+- The compiler knows which a flags are used in the Responding and which are used in the Update phases
+  That information can be used to avoid unnecessary calls to update() or p()
+  (Do we want to call beforeUpdate and afterUpdate if p() isn't going to do anything?)
+- Compiler knows the names corresponding with the flags, exposing those names on the component in dev mode would allow better error messages. "`name` was changed too many times"
+
+## How we teach this
+
+The improvements don't require changes of existing svelte code, the reactivity remains invisible, and for new users comes with less caveats.
+
+If code depends on reactivity not working, that code will need to be updated.
+
+Some improvements like error recovery will need to be added to the docs.
+
+## Drawbacks
+
+The current implementation will be faster but I favor correctness over speed.  
+(reactive statements are run 0 or 1x per `$$.update()`)
+
+The proposed system is still not 100% correct.
+
+## Alternatives
+
+### Detect loop by mask patterns
+
+Detecting a looping patterns pattern in the masks:
+
+```
+update 1: changed A & B
+update 2: changed C & A
+update 3: changed D
+update 4: changed A & B
+update 5: changed C & A
+update 6: changed D
+```
+
+update 1 to 3 is the same as 4 - 6, this is probably a loop.
+
+Pro: Leads
+Cons: Slower, more memory usage, difficult to implement, still not 100% correct
+
+### Diffing the mask
+
+In the reactive phase: A call to `$$invalidate()` should add a diry flag to a separate dirty mask but don't update the mask the `$$.update()` is working on. After the update ran, compare the dirty mask to the mask the `$$.update()` started with and call `$$.update()` again with only the flags that were missing before. This solves the bug of not running a handler at all based on ordering, but doesn't solve correctness, if a handler is called with the latest or an outdated value still depends on the ordering.
+
+Pro's: No looping issues, no duplicate calls to handlers  
+Cons: Not predictable
+
+### Limit update to run n times
+
+Run the `$$.update()` until no `$$invalidate` is called or a limit of say 100 is reached.
+If the limit is reached either show a warning in the console and continue or silently start the next phase.
+
+Pro: Easy to implement  
+Cons: How many should n be? creating a computed property based on another computed property is allowed, but now suddenly has a limit.  
+Creates slow downs when a component creates a loop.
+
+### No protection against infinite loop
```

Comment:
@bfanger thanks for this writeup. very educational. I recently created [this issue](https://github.com/sveltejs/svelte/issues/6732) and would really like to add the ability to opt out of infinite loop protection because it is breaking some of our production code in unexpected ways. I started reading through the code but obviously your understanding far exceeds mine. How would you go about disabling it in a proper way? thanks!